File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"io"
	"net"
)

const (
	OpcodeBet   byte = 0x01
	OpcodeAck   byte = 0x02
	OpcodeError byte = 0x03
	OpcodeBatch byte = 0x04
	OpcodeEndData    byte = 0x05
    OpcodeGetWinners byte = 0x06
)

type Frame struct {
	Opcode byte
	Body   []byte
}

// WriteFrame sends the header (5 bytes) + body over the network
func WriteFrame(conn net.Conn, opcode byte, body []byte) error {
	header := make([]byte, 5)
	header[0] = opcode
	binary.BigEndian.PutUint32(header[1:], uint32(len(body)))
	    data := append(header, body...)
    totalSent := 0

    for totalSent < len(data) {
        n, err := conn.Write(data[totalSent:])
        if err != nil {
            return err
        }
        if n == 0 {
            break
        }
        totalSent += n
    }

    return nil
}

// ReadFrame reads exactly 5 bytes for the header, then 'length' bytes for the body
func ReadFrame(conn net.Conn) (*Frame, error) {
	header := make([]byte, 5)
	if _, err := io.ReadFull(conn, header); err != nil {
		return nil, err
	}

	opcode := header[0]
	length := binary.BigEndian.Uint32(header[1:])

	body := make([]byte, length)
	if _, err := io.ReadFull(conn, body); err != nil {
		return nil, err
	}

	return &Frame{Opcode: opcode, Body: body}, nil
}

File: server/common/protocol.py
import struct

OPCODE_BET = 0x01
OPCODE_ACK = 0x02
OPCODE_ERROR = 0x03
OPCODE_BATCH = 0x04
OPCODE_END_DATA = 0x05
OPCODE_GET_WINNERS = 0x06

HEADER_SIZE = 5

class Protocol:
    @staticmethod
    def send_frame(sock, opcode, body):
        # Header: 1 byte opcode + 4 bytes length (Big Endian '>')
        header = struct.pack('>BI', opcode, len(body))
        sock.sendall(header + body)

    @staticmethod
    def receive_frame(sock):
        # 1. Read Header
        header_raw = Protocol._recv_all(sock, HEADER_SIZE)
        if not header_raw: return None
        
        opcode, length = struct.unpack('>BI', header_raw)
        
        # 2. Read Body
        body = Protocol._recv_all(sock, length)
        return opcode, body

    @staticmethod
    def _recv_all(sock, n):
        """ Helper to handle Short Reads: keeps reading until n bytes are received """
        data = bytearray()
        while len(data) < n:
            packet = sock.recv(n - len(data))
            if not packet:
                return None # Connection closed
            data.extend(packet)
        return bytes(data)